Content.Client/_ES/Viewcone/Overlays/ESViewconeSetAlphaOverlay.cs

```diff
@@ -0,0 +1,129 @@
+using Content.Client._ES.Viewcone.ComponentTree;
+using Content.Shared._ES.Viewcone;
+using Content.Shared.MouseRotator;
+using Robust.Client.GameObjects;
+using Robust.Client.Graphics;
+using Robust.Client.Input;
+using Robust.Shared.Enums;
+using Robust.Shared.Map.Components;
+
+namespace Content.Client._ES.Viewcone.Overlays;
+
+/// <summary>
+///     Queries the bounds for each viewport for all <see cref="ESViewconeOccludableComponent"/>, then
+///     sets their alpha before entities render in accordance with whether they should be in view or not
+///
+///     This alpha pass only works because of <see cref="ESViewconeResetAlphaOverlay"/>, which resets in a later stage of rendering.
+/// </summary>
+public sealed class ESViewconeSetAlphaOverlay : Overlay
+{
+    [Dependency] private readonly IEntityManager _ent = default!;
+    [Dependency] private readonly IEyeManager _eye = default!;
+    [Dependency] private readonly IInputManager _input = default!;
+    private readonly ESViewconeOverlayManagementSystem _cone;
+    private readonly ESViewconeOccludableTreeSystem _tree;
+    private readonly TransformSystem _xform;
+    private readonly SpriteSystem _sprite;
+
+    public override OverlaySpace Space => OverlaySpace.WorldSpaceBelowEntities;
+
+    // slightly sus but cached from beforedraw to use in draw.
+    private Entity<EyeComponent, ESViewconeComponent>? _nextEye;
+
+    public ESViewconeSetAlphaOverlay()
+    {
+        IoCManager.InjectDependencies(this);
+
+        _cone = _ent.EntitySysManager.GetEntitySystem<ESViewconeOverlayManagementSystem>();
+        _tree = _ent.EntitySysManager.GetEntitySystem<ESViewconeOccludableTreeSystem>();
+        _xform  = _ent.EntitySysManager.GetEntitySystem<TransformSystem>();
+        _sprite = _ent.EntitySysManager.GetEntitySystem<SpriteSystem>();
+    }
+
+    protected override bool BeforeDraw(in OverlayDrawArgs args)
+    {
+        _nextEye = null;
+
+        if (args.Viewport.Eye == null)
+            return false;
+
+        // This is really stupid but there isn't another way to reverse an eye entity from just an IEye afaict
+        // It's not really inefficient though. theres barely any of those fuckin things anyway (? verify that) (maybe this scales with players in view) (shit)
+        var enumerator = _ent.AllEntityQueryEnumerator<EyeComponent, ESViewconeComponent>();
+        while (enumerator.MoveNext(out var uid, out var eye, out var viewcone))
+        {
+            if (args.Viewport.Eye != eye.Eye)
+                continue;
+
+            _nextEye = (uid, eye, viewcone);
+            break;
+        }
+
+        return _nextEye != null;
```

Comment:
realistically the most sus thing here. couldnt think of a better way to go from an `IEye` to the `EyeComponent` that holds it without engine changes. but also like this is completely fine